### PR TITLE
[Draft] Nemotron in Nemo-UX

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        super().__init__(config or NemotronConfig(), optim=optim, tokenizer=tokenizer, model_transform=model_transform)
+
+
+@io.model_importer(NemotronModel, "hf")
+class HFNemotronImporter(io.ModelConnector["NemotronForCausalLM", NemotronModel]):
+    def init(self) -> NemotronModel:
+        return NemotronModel(self.config, tokenizer=self.tokenizer)
+
+    def apply(self, output_path: Path) -> Path:
+        from transformers import NemotronForCausalLM
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:95]: Imported at top of the file

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+    @property
+    def tokenizer(self) -> "AutoTokenizer":
+        from nemo.collections.common.tokenizers.huggingface.auto_tokenizer import AutoTokenizer
+
+        return AutoTokenizer(str(self))
+
+    @property
+    def config(self) -> NemotronConfig:
+        from transformers import NemotronConfig as HFNemotronConfig
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:135]: Imported at top of the file

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+    @property
+    def tokenizer(self):
+        return io.load_context(str(self)).model.tokenizer.tokenizer
+
+    @property
+    def config(self) -> "HFNemotronConfig":
+        source: NemotronConfig = io.load_context(str(self)).model.config
+
+        from transformers import NemotronConfig as HFNemotronConfig
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:206]: Imported at top of the file

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            share_embeddings_and_output_weights=False,
+        )
+
+        return output
+
+
+@io.model_exporter(NemotronModel, "hf")
+class HFNemotronExporter(io.ModelConnector[NemotronModel, "NemotronForCausalLM"]):
+    def init(self) -> "NemotronForCausalLM":
+        from transformers import AutoModelForCausalLM
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:166]: Can we import it at top of the file with the other transformers modules ?

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    v_proj = linear_qkv[v_slice].reshape(-1, hidden_size).cpu()
+
+    return q_proj, k_proj, v_proj
+
+
+__all__ = [
+    "NemotronConfig",
+    "Nemotron3Config4B",
+    "Nemotron3Config8B",
+    "Nemotron4Config15B",
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:311]: Missing 22B model

File: nemo/collections/llm/gpt/model/nemotron.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+if TYPE_CHECKING:
+    from transformers import NemotronConfig as HFNemotronConfig
+    from transformers import NemotronForCausalLM
+
+    from nemo.collections.common.tokenizers.huggingface.auto_tokenizer import AutoTokenizer
+    from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+
+
+@dataclass
```
Comment (ahmadki) [inline nemo/collections/llm/gpt/model/nemotron.py:21]: This is mostly a personal preference, but I prefer the following structure to configs:

```
@dataclass
class NemotronConfig(GPTConfig):
    # configs that are common across model sizes
    normalization: str = "LayerNorm"
    activation_func: Callable = squared_relu
    position_embedding_type: str = "rope"
    share_embeddings_and_output_weights: bool = False
    add_bias_linear: bool = False

    hidden_dropout: float = 0.0
    attention_dropout: float = 0.0
    apply_query_key_layer_scaling: bool = True
    rotary_percent: float = 0.5
    masked_softmax_fusion: bool = True
    persist_layer_norm: bool = True
    bias_dropout_add_fusion: bool = False
    layernorm_zero_centered_gamma: bool = True

    # Nemotron3Config4B as default configs
    num_layers: int = 32
    seq_length: int = 4096
    hidden_size: int = 3072
    ffn_hidden_size: int = 9216
    num_attention_heads: int = 24
    num_query_groups: Optional[int] = 8
    kv_channels: Optional[int] = 128
    init_method_std: float = 0.0134

@dataclass
class Nemotron3Config4B(NemotronConfig):
    num_layers: int = 32
    seq_length: int = 4096
    hidden_size: int = 3072
    ffn_hidden_size: int = 9216
    num_attention_heads: int = 24
    num_query_groups: int = 8
    kv_channels: Optional[int] = 128
    init_method_std: float = 0.0134


@dataclass
class Nemotron3Config8B(NemotronConfig):
    num_layers: int = 32
    seq_length: int = 4096
    hidden_size: int = 4096
    ffn_hidden_size: int = 16384
    num_attention_heads: int = 32
    num_query_groups: Optional[int] = None
    kv_channels: Optional[int] = None
    init_method_std: float = 0.010


@dataclass
class Nemotron4Config15B(NemotronConfig):
    num_layers: int = 32
    seq_length: int = 4096
    hidden_size: int = 6144
    ffn_hidden_size: int = 24576
    num_attention_heads: int = 48
    num_query_groups: Optional[int] = 8
    kv_channels: Optional[int] = None
    init_method_std: float = 0.0134


@dataclass
class Nemotron4Config22B(NemotronConfig):
    num_layers: int = 40
    seq_length: int = 4096
    hidden_size: int = 6144
    ffn_hidden_size: int = 24576
    num_attention_heads: int = 48
    num_query_groups: Optional[int] = None
    kv_channels: Optional[int] = None
    init_method_std: float = 0.008


@dataclass
class Nemotron4Config340B(NemotronConfig):
    num_layers: int = 96
    seq_length: int = 4096
    hidden_size: int = 18432
    ffn_hidden_size: int = 73728
    num_attention_heads: int = 96
    num_query_groups: Optional[int] = 8
    kv_channels: Optional[int] = None
    init_method_std: float = 0.0063
```

1. `NemotronConfig` should be usable as a valid nemotron config by it self. So I give it the 4b model specs by default even if we duplicate them in `Nemotron3Config4B `
2. it is easy to identify common values between configs, and version specific values.
3. if one config overrides a value from `NemotronConfig`, then all configs should have this value clearly set. It makes it easier to identify differences between models.
4. configs that take `None` values should be `Optional`.
5. missing `init_method_std ` for 8b
6. some values are missing from `NemotronConfig`,  I'm not sure how they affect bit-exactness:
```
apply_query_key_layer_scaling: bool = True
masked_softmax_fusion: bool = True
persist_layer_norm: bool = True
bias_dropout_add_fusion: bool = False
```

Comment (suiyoubi) [reply nemo/collections/llm/gpt/model/nemotron.py:20]: sure that makes sense. I'll adapt to this

**Review by ahmadki — APPROVED** [review]

**Pull request merged**